tests/python_package_test/test_dask.py

```diff
@@ -255,7 +255,7 @@ def test_classifier(output, task, boosting_type, tree_learner, client):
             'bagging_fraction': 0.9,
         })
     elif boosting_type == 'goss':
-        params['top_rate'] = 0.5
+        params['top_rate'] = 0.7
```

Comment:
it looks like this was added since I last reviewed (https://github.com/microsoft/LightGBM/pull/4149/commits/981084f5360b55daae97fe54d0679e0a12fc1703). Can you please explain why it's necessary?

Reply:
`test_classifier` became flaky in this PR. I assume it's because previously we weren't performing distributed training or at least not everytime, so adding this generated some fails in multiclass classification for [data_parallel-dart](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=9683&view=logs&j=275189f9-c769-596a-7ef9-49fb48a9ab70&t=3a9e7a4a-04e6-52a0-67ea-6e8f6cfda74f&l=533), [voting_parallel-rf](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=9683&view=logs&j=275189f9-c769-596a-7ef9-49fb48a9ab70&t=3a9e7a4a-04e6-52a0-67ea-6e8f6cfda74f&l=779) (this one is very surprising, given that the atol is 0.8), [voting_parallel-gbdt](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=9683&view=logs&j=e3619e06-e539-5956-1ae7-e8c2c29d8e91&t=4b3d6e5d-f8c3-577f-235f-b68c5ed4d3ab&l=1569), [voting_parallel-dart](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=9683&view=logs&j=e3619e06-e539-5956-1ae7-e8c2c29d8e91&t=4b3d6e5d-f8c3-577f-235f-b68c5ed4d3ab&l=1800), [voting_parallel-goss](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=9683&view=logs&j=b463fb3a-b487-5cfc-fc07-6d216464ba86&t=c16b9cdb-c4c8-59f9-2ae2-b81f62c3a9d0&l=547). Most of them are for dataframe with categoricals but there are a couple with sparse matrices. I have to debug them to see what's actually happening, this is a very simple classification problem and I'd expect to get a perfect score with little effort. I'll ping you here once I'm done but it could take a bit haha.

Reply:
got it, thanks! Let me know if you need any help

Reply:
> I'd expect to get a perfect score with little effort

Given the small dataset sizes we use in tests, I think it would be useful to set  `min_data_in_leaf: 0` everywhere. That might improve the predictability of the results.

Reply:
Sorry this is taking so long, I haven't had much time and I'm really confused by this. The same data point makes the test fail even for `data_parallel` and `gbdt`, I'm trying to figure out what's exactly going on here, I have the test in a while loop and it eventually fails because of that data point, I'm not sure what's wrong with it haha.

Reply:
Btw, setting `min_data_in_leaf=0` gives this error: `LightGBMError: Check failed: (best_split_info.right_count) > (0) at /hdd/github/LightGBM/src/treelearner/serial_tree_learner.cpp, line 663`. Do you think this could be related to #4026? This data is shuffled but I think forcing few samples in a leaf gives more chance to getting an empty split in one of the workers.

Reply:
Here's `abs(local_probas - dask_probas)` per iteration for `data_parallel` `gbdt` for just that one sample (index 377):
![image](https://user-images.githubusercontent.com/8473587/114493787-c67d1d80-9be0-11eb-8369-5e47b171caed.png)
So from the 7th iteration onwards the probabilities start to increasingly differ, I think there's definitely something strange going on here.